### ISSUE TITLE
[JENKINS-73] CVS changelog always shows edited files as added

### ISSUE THREAD

**Issue by jenkins-infra-bot** [timja-org/jenkins-gh-issues-poc-11-14#2071]
<p>All modified files (trunk or branch) in a CVS changes list are shown as added,<br/>
even when they really were modified. And as far as I can tell, files added on<br/>
branches are shown as edits. Files added in trunk appear correctly as adds.<br/>
Deletions in trunk also seem to be shown as adds.</p>

<p>The problem seems to be in the change log generator, not Hudson's parser, since<br/>
changelog.xml really shows things like:</p>

<p><entry><br/>
  <date>...</date><br/>
  <time>...</time><br/>
  <author><![CDATA<span class="error">&#91;...&#93;</span>]></author><br/>
  <file><br/>
    <name>.../X.java</name><br/>
    <revision>1.1.2.9</revision><br/>
  </file><br/>
  <msg><![CDATA<span class="error">&#91;...an edit; similar for a trunk commit...&#93;</span>]></msg><br/>
</entry></p>

<p>(note no <prevrevision>1.1.2.8</prevrevision>) or</p>

<p><entry><br/>
  <date>...</date><br/>
  <time>...</time><br/>
  <author><![CDATA<span class="error">&#91;...&#93;</span>]></author><br/>
  <file><br/>
    <name>.../X.java</name><br/>
    <revision>1.1.2.1</revision><br/>
  </file><br/>
  <msg><![CDATA<span class="error">&#91;...an add...&#93;</span>]></msg><br/>
</entry><br/>
<entry><br/>
  <date>...same...</date><br/>
  <time>...same...</time><br/>
  <author><![CDATA<span class="error">&#91;...same...&#93;</span>]></author><br/>
  <file><br/>
    <name>.../X.java</name><br/>
    <revision>1.1</revision><br/>
    <prevrevision>1.1.2.1</prevrevision><br/>
  </file><br/>
  <msg><![CDATA[branches:  1.1.2;<br/>
file X.java was initially added on branch xxx.]]></msg><br/>
</entry></p>

<p>(note synthetic add message, and <revision> vs. <prevrevision> backwards)</p>

<p><entry><br/>
  <date>...</date><br/>
  <time>...</time><br/>
  <author><![CDATA<span class="error">&#91;...&#93;</span>]></author><br/>
  <file><br/>
    <name>.../X.java</name><br/>
    <revision>1.2</revision><br/>
  </file><br/>
  <msg><![CDATA<span class="error">&#91;...useless now...&#93;</span>]></msg><br/>
</entry></p>

<p>(note no metadata indicating this was a deletion of file rev 1.1)</p>

---
<details><summary><i>Originally reported by <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick">jglick</a>, imported from: <a href="https://issues.jenkins.io/browse/JENKINS-73" target="_blank">CVS changelog always shows edited files as added</a></i></summary>
<i><ul>
<li><b>status</b>: Closed
<li><b>priority</b>: Minor
<li><b>resolution</b>: Postponed
<li><b>resolved</b>: 2006-07-28T21:40:31+00:00
<li><b>votes</b>: 0
<li><b>watchers</b>: 1
<li><b>imported</b>: 2025-11-14
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>All modified files (trunk or branch) in a CVS changes list are shown as added,
even when they really were modified. And as far as I can tell, files added on
branches are shown as edits. Files added in trunk appear correctly as adds.
Deletions in trunk also seem to be shown as adds.</p>

<p>The problem seems to be in the change log generator, not Hudson's parser, since
changelog.xml really shows things like:</p>

<p>&lt;entry&gt;
  &lt;date&gt;...&lt;/date&gt;
  &lt;time&gt;...&lt;/time&gt;
  &lt;author&gt;&lt;![CDATA<span class="error">&#91;...&#93;</span>]&gt;&lt;/author&gt;
  &lt;file&gt;
    &lt;name&gt;.../X.java&lt;/name&gt;
    &lt;revision&gt;1.1.2.9&lt;/revision&gt;
  &lt;/file&gt;
  &lt;msg&gt;&lt;![CDATA<span class="error">&#91;...an edit; similar for a trunk commit...&#93;</span>]&gt;&lt;/msg&gt;
&lt;/entry&gt;</p>

<p>(note no &lt;prevrevision&gt;1.1.2.8&lt;/prevrevision&gt;) or</p>

<p>&lt;entry&gt;
  &lt;date&gt;...&lt;/date&gt;
  &lt;time&gt;...&lt;/time&gt;
  &lt;author&gt;&lt;![CDATA<span class="error">&#91;...&#93;</span>]&gt;&lt;/author&gt;
  &lt;file&gt;
    &lt;name&gt;.../X.java&lt;/name&gt;
    &lt;revision&gt;1.1.2.1&lt;/revision&gt;
  &lt;/file&gt;
  &lt;msg&gt;&lt;![CDATA<span class="error">&#91;...an add...&#93;</span>]&gt;&lt;/msg&gt;
&lt;/entry&gt;
&lt;entry&gt;
  &lt;date&gt;...same...&lt;/date&gt;
  &lt;time&gt;...same...&lt;/time&gt;
  &lt;author&gt;&lt;![CDATA<span class="error">&#91;...same...&#93;</span>]&gt;&lt;/author&gt;
  &lt;file&gt;
    &lt;name&gt;.../X.java&lt;/name&gt;
    &lt;revision&gt;1.1&lt;/revision&gt;
    &lt;prevrevision&gt;1.1.2.1&lt;/prevrevision&gt;
  &lt;/file&gt;
  &lt;msg&gt;&lt;![CDATA[branches:  1.1.2;
file X.java was initially added on branch xxx.]]&gt;&lt;/msg&gt;
&lt;/entry&gt;</p>

<p>(note synthetic add message, and &lt;revision&gt; vs. &lt;prevrevision&gt; backwards)</p>

<p>&lt;entry&gt;
  &lt;date&gt;...&lt;/date&gt;
  &lt;time&gt;...&lt;/time&gt;
  &lt;author&gt;&lt;![CDATA<span class="error">&#91;...&#93;</span>]&gt;&lt;/author&gt;
  &lt;file&gt;
    &lt;name&gt;.../X.java&lt;/name&gt;
    &lt;revision&gt;1.2&lt;/revision&gt;
  &lt;/file&gt;
  &lt;msg&gt;&lt;![CDATA<span class="error">&#91;...useless now...&#93;</span>]&gt;&lt;/msg&gt;
&lt;/entry&gt;</p>

<p>(note no metadata indicating this was a deletion of file rev 1.1)</p></pre>
</details>
</details>
<ul><li><i>environment</i>: <code>Platform: All, OS: All</code></li></ul>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kohsuke.png" title="kohsuke's avatar" /> kohsuke's <a href="https://issues.jenkins.io/browse/JENKINS-73?focusedId=119842&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-119842">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>CVS log output looks like the following:</p>

<p>----------------------------
revision 1.16
date: 2006/07/26 21:00:10;  author: kohsuke;  state: Exp;  lines: +1 -0;  kopt:
kv;  commitid: 147c44c7d7da4eb2;  filename: foo.txt;
changed
----------------------------
revision 1.15
date: 2006/07/26 20:57:00;  author: kohsuke;  state: Exp;  lines: +1 -0;  kopt:
kv;  commitid: 150044c7d71c4c46;  filename: foo.txt;
changed
----------------------------
revision 1.14
date: 2006/07/26 20:55:37;  author: kohsuke;  state: Exp;  lines: +1 -0;  kopt:
kv;  commitid: 166c44c7d6c84b34;  filename: foo.txt;
changed
=============================================================================</p>

<p>And '&lt;prevrevision&gt;' is obtained only when the "cvs log" command fetched more
than one entries from the server. The revision of the previous log record is
used as the 'previous revision' of the current log record.</p>

<p>Since Hudson runs "cvs log" with -d option, chances are, it only selects log
records that happened since the last build, and in that case, changelog.xml
doesn't record &lt;prevrevision&gt; because CVS didn't return any.</p>

<p>Therefore, it seems like after all it's not always possible to determine
add/edit/remove just from the CVS log command. We can distinguish remove from
other two by looking for "state: dead" line, but we can't distinguish add and
edit unless we check the previous revision.</p>

<p>I thought about using the output from "cvs update", but I still don't think we
can distinguish add and edit. I believe both may produce 'U' record.</p>

<p>Perhaps doing this correctly depends on issue #49. Then we can reliably find out
what data we are exchanging with the CVS server.</p>
</pre>
</details>

</details>
<p>CVS log output looks like the following:</p>

<p>----------------------------<br/>
revision 1.16<br/>
date: 2006/07/26 21:00:10;  author: kohsuke;  state: Exp;  lines: +1 -0;  kopt:<br/>
kv;  commitid: 147c44c7d7da4eb2;  filename: foo.txt;<br/>
changed<br/>
----------------------------<br/>
revision 1.15<br/>
date: 2006/07/26 20:57:00;  author: kohsuke;  state: Exp;  lines: +1 -0;  kopt:<br/>
kv;  commitid: 150044c7d71c4c46;  filename: foo.txt;<br/>
changed<br/>
----------------------------<br/>
revision 1.14<br/>
date: 2006/07/26 20:55:37;  author: kohsuke;  state: Exp;  lines: +1 -0;  kopt:<br/>
kv;  commitid: 166c44c7d6c84b34;  filename: foo.txt;<br/>
changed<br/>
=============================================================================</p>

<p>And '<prevrevision>' is obtained only when the "cvs log" command fetched more<br/>
than one entries from the server. The revision of the previous log record is<br/>
used as the 'previous revision' of the current log record.</p>

<p>Since Hudson runs "cvs log" with -d option, chances are, it only selects log<br/>
records that happened since the last build, and in that case, changelog.xml<br/>
doesn't record <prevrevision> because CVS didn't return any.</p>

<p>Therefore, it seems like after all it's not always possible to determine<br/>
add/edit/remove just from the CVS log command. We can distinguish remove from<br/>
other two by looking for "state: dead" line, but we can't distinguish add and<br/>
edit unless we check the previous revision.</p>

<p>I thought about using the output from "cvs update", but I still don't think we<br/>
can distinguish add and edit. I believe both may produce 'U' record.</p>

<p>Perhaps doing this correctly depends on issue #49. Then we can reliably find out<br/>
what data we are exchanging with the CVS server.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kohsuke.png" title="kohsuke's avatar" /> kohsuke's <a href="https://issues.jenkins.io/browse/JENKINS-73?focusedId=119843&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-119843">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>Meanwhile, I improved the parser to at least detect deletion.</p>
</pre>
</details>

</details>
<p>Meanwhile, I improved the parser to at least detect deletion.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> jglick's <a href="https://issues.jenkins.io/browse/JENKINS-73?focusedId=119844&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-119844">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>Add vs. Edit is tricky because you can have a file 1.1 deleted in 1.2 and
readded in 1.3, so just checking rev.equals("1.1") will not always work. Also,
adding on a branch creates 1.1 as DEAD and a real rev 1.1.${2*k}.1 for some k&gt;0,
and I presume readding a dead file on a branch produces some other form still.
So without a full file history it is hard to do more than guess. Since Hudson
now only asks the server for logs for files which really changed in the last
update, maybe it could leave off the -d arg to cvs log and try to extract the
important bits itself. Ugh.</p>

<p>Anyway I consider this a relatively low priority issue; it took me a long time
to even notice that document_add.gif was supposed to mean "added file". The file
name and log message are the important bits - if you really want to know what
changed you had better go find the diff, e.g. subscribe to the CVS change list
or have a ViewCVS server installed somewhere.</p>
</pre>
</details>

</details>
<p>Add vs. Edit is tricky because you can have a file 1.1 deleted in 1.2 and<br/>
readded in 1.3, so just checking rev.equals("1.1") will not always work. Also,<br/>
adding on a branch creates 1.1 as DEAD and a real rev 1.1.${2*k}.1 for some k>0,<br/>
and I presume readding a dead file on a branch produces some other form still.<br/>
So without a full file history it is hard to do more than guess. Since Hudson<br/>
now only asks the server for logs for files which really changed in the last<br/>
update, maybe it could leave off the -d arg to cvs log and try to extract the<br/>
important bits itself. Ugh.</p>

<p>Anyway I consider this a relatively low priority issue; it took me a long time<br/>
to even notice that document_add.gif was supposed to mean "added file". The file<br/>
name and log message are the important bits - if you really want to know what<br/>
changed you had better go find the diff, e.g. subscribe to the CVS change list<br/>
or have a ViewCVS server installed somewhere.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kohsuke.png" title="kohsuke's avatar" /> kohsuke's <a href="https://issues.jenkins.io/browse/JENKINS-73?focusedId=119845&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-119845">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>I fixed "1.1" handling. If we know "1.1" is not remove, then it's always add,
because it cannot be edit.</p>

<p>The rest has to wait for 49. Like you say, file names and comments are the
important bits.</p>
</pre>
</details>

</details>
<p>I fixed "1.1" handling. If we know "1.1" is not remove, then it's always add,<br/>
because it cannot be edit.</p>

<p>The rest has to wait for 49. Like you say, file names and comments are the<br/>
important bits.</p>

**Comment by jenkins-infra-bot** [comment]
<i>[Originally depends on: <a href="https://github.com/timja-org/jenkins-gh-issues-poc-11-14/issues?q=in%3Atitle%20JENKINS-49">JENKINS-49</a>]</i>
